### PR TITLE
Implemented 'focusIn,focusOut' events emulation (close  #1044)

### LINKED ISSUE #1044: Emulate focusin and focusout events in background
When a window is in background, browser stops raising focus events even after `.focus()` call. Currently in this case we manually raise `focus` and `blur` events. We need to do  the same for `focusin` and `focusout` events.

### PR DESCRIPTION
**Pull request by miherlosev**

### CODE REVIEW

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
FPR
@helen-dikareva @AndreyBelym

File: src/client/sandbox/event/focus-blur.js [hunk 1]
```diff
@@ -64,85 +63,117 @@ export default class FocusBlurSandbox extends SandboxBase {
...
-            FocusBlurSandbox._getNativeMeth(el, type).call(el);
-
-            if (preventScrolling)
-                FocusBlurSandbox._restoreElementScroll(this.window, windowScroll);
-
-            if (browserUtils.isIE && nonScrollableParents.length) {
-                for (var j = 0, len = nonScrollableParents.length; j < len; j++)
-                    FocusBlurSandbox._restoreElementScroll(nonScrollableParents[j], nonScrollableParentsScrollValues[j]);
-            }
+            // NOTE: To guarantee that all focus/blur events are raised, we need to raise them manually.
```
Comment (helen-dikareva) [inline src/client/sandbox/event/focus-blur.js]: This comment is not related with scrolling. 
I don't really like that `_ensureWindowScroll` call real `focus`.

File: src/client/sandbox/event/focus-blur.js [hunk 2]
```diff
@@ -64,85 +63,117 @@ export default class FocusBlurSandbox extends SandboxBase {
...
+            }
+        }
+    }
+
+    _getElementNonScrollableParentsScrollState (el) {
+        var scrollState    = [];
+        var elementParents = domUtils.getParents(el);
+        var elementParent  = null;
+
+        for (var i = 0; i < elementParents.length; i++) {
```
Comment (helen-dikareva) [inline src/client/sandbox/event/focus-blur.js]: You can use `for ... of` in circles

File: src/client/sandbox/event/focus-blur.js [hunk 3]
```diff
@@ -172,7 +203,25 @@ export default class FocusBlurSandbox extends SandboxBase {
...
+        else if (focusInOutEventType === 'focusout')
+            return 'blur';
+
+        return null;
+    }
+
+    static getFocusInOutEventType (focusBlurEventType) {
+        if (focusBlurEventType === 'focus')
+            return 'focusin';
+        else if (focusBlurEventType === 'blur')
```
Comment (helen-dikareva) [inline src/client/sandbox/event/focus-blur.js]: `getFocusBlurEventType`
`getFocusInOutEventType`
These two methods look very similar and confusing

File: test/client/fixtures/sandbox/event/focus-blur-change-test.js [hunk 4]
```diff
@@ -879,6 +879,71 @@ asyncTest("focus() must not scroll to the element if 'preventScrolling' argument
...
+        var nativeEventLog = '';
+        var input          = document.createElement('input');
+        var nativeInput    = nativeMethods.createElement.call(document, 'input');
+
+        document.body.appendChild(input);
+        nativeInput.focus = nativeMethods.focus;
+        nativeInput.blur  = nativeMethods.blur;
+        nativeMethods.appendChild.call(document.body, nativeInput);
+
+        var focusHandler                    = function () {
```
Comment (helen-dikareva) [inline test/client/fixtures/sandbox/event/focus-blur-change-test.js]: It can be one handler:
```js
function (e) {
    eventLog +=  e.type + '|';
};
```

Comment (miherlosev) [inline test/client/fixtures/sandbox/event/focus-blur-change-test.js]: Ok

File: src/client/sandbox/event/focus-blur.js (hunk 3 above)
Comment (miherlosev) [inline src/client/sandbox/event/focus-blur.js]: > redundant else

Ok

File: src/client/sandbox/event/focus-blur.js (hunk 1 above)
Comment (miherlosev) [inline src/client/sandbox/event/focus-blur.js]: > This comment is not related with scrolling. 

This comment is placed on first line in brackets in which  we perform `_raiseEvent` calls.
https://github.com/miherlosev/testcafe-hammerhead/blob/74b2da15b2634991b3f66d762b90946d5d1e4ec6/src/client/sandbox/event/focus-blur.js#L175

File: src/client/sandbox/event/focus-blur.js (hunk 2 above)
Comment (miherlosev) [inline src/client/sandbox/event/focus-blur.js]: Ok

File: src/client/sandbox/event/focus-blur.js (hunk 3 above)
Comment (miherlosev) [inline src/client/sandbox/event/focus-blur.js]: >  getFocusBlurEventType
>  getFocusInOutEventType
>  These two methods look very similar and confusing

I need to perform two way convesion:
`blur <-> focusout`
`focus <-> focusin`

Propose a better way by your opinion

File: src/client/sandbox/event/focus-blur.js (hunk 1 above)
Comment (miherlosev) [inline src/client/sandbox/event/focus-blur.js]: >  I don't really like that _ensureWindowScroll call real focus

See the current code from
https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/client/sandbox/event/focus-blur.js#L104
to 
https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/client/sandbox/event/focus-blur.js#L145
I don't change an old logic
I just split the large code part on simple logical pars

**Comment by miherlosev** [comment]
FPR

**Comment by miherlosev** [comment]
FPR @helen-dikareva

File: src/client/sandbox/event/focus-blur.js [hunk 5]
```diff
@@ -181,7 +148,36 @@ export default class FocusBlurSandbox extends SandboxBase {
...
+            let windowScroll                           = null;
+            let elementNonScrollableParentsScrollState = null;
+
+            if (preventScrolling)
+                windowScroll = styleUtils.getElementScroll(this.window);
+
+            if (browserUtils.isIE)
+                elementNonScrollableParentsScrollState = this._getElementNonScrollableParentsScrollState(el);
+
+            if (this._shouldUseLabelHtmlForElement(el, type)) {
```
Comment (helen-dikareva) [inline src/client/sandbox/event/focus-blur.js]: I think it was better with ` this._ensureWindowScroll(el, preventScrolling, type, callback);`. Only without `callback`
What about
```js
el =  this._ensureWindowScroll(el, preventScrolling, type);

if (!el) {
    callback();
    return;
}

File: src/client/sandbox/event/focus-blur.js [hunk 6]
```diff
@@ -181,7 +148,36 @@ export default class FocusBlurSandbox extends SandboxBase {
 
         if (el[type]) {
             // NOTE: To guarantee that all focus/blur events are raised, we need to raise them manually.
-            this._ensureWindowScroll(el, preventScrolling, type, callback);
+            let windowScroll                           = null;
+            let elementNonScrollableParentsScrollState = null;
+
+            if (preventScrolling)
```
Comment (helen-dikareva) [inline src/client/sandbox/event/focus-blur.js]: May be we can extract in method storing scroll state `saveScrollStateIfNecessary` and for storing valuesuse` this`.
And after calling `focus` use method` restoreScrollStateIfNecessary`

Comment (churkin) [inline src/client/sandbox/event/focus-blur.js]: Agree

File: src/client/sandbox/event/focus-blur.js (hunk 5 above)
Comment (miherlosev) [inline src/client/sandbox/event/focus-blur.js]: I don't like this way.
It is not clear why `_ensureWindowScroll` method retruns `el` or `null`.

Comment (helen-dikareva) [inline src/client/sandbox/event/focus-blur.js]: I'm sorry `_ensureLabelHtmlForElement` of course

Comment (miherlosev) [inline src/client/sandbox/event/focus-blur.js]: It is a bad solution too.
You should returns two fields: `el` and flag that equals `it is  label with htmlFor attribute`.
Then we will return is only if it is label with htmlFor attribute and element is null.
So, the current way is simpler.

Comment (helen-dikareva) [inline src/client/sandbox/event/focus-blur.js]: We don't need any flags. We should just return `el`. What do you mean?

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
Tests are green.
@helen-dikareva merge this
